### PR TITLE
Adding export helper. Solves #5633

### PR DESCRIPTION
**Pull request by Troush**
- [x]  `e/1` function that list functions and macros in the same way the autocomplete.
- [x]  make `module_info_funs/1` public
- [x]  rename `module_info_funs/1` to `exports/1`
- [x]  tests

### CODE REVIEW

**Comment by riseshia** [comment]
Related to #5633

**Comment by josevalim** [comment]
This looks great, thank you. The only change I would do is to generalize the `ls_print` into `table_print` or something of sorts. It will be quite similar to `ls_print` today except the `format_item+Path.join` will be passed as as anonymous function. This way we won't call Path.join nor format_item when formatting the exports:

https://github.com/Troush/elixir/blob/99baf3161ae091a95e1a0757f7aea8c7cc7655a4/lib/iex/lib/iex/helpers.ex#L596

This would also open up the possibility of formatting functions and macros differently if we want to in the future. :)

**Comment by Troush** [comment]
I took the liberty to change the name of `ls_print` to `tp` and make it public as helper for cases as: ```tp(Enum.map(1..10, &Integer.to_string/1))```

**Comment by josevalim** [comment]
Let's please not make `tp` public for now. We can always do so later but we would need to discuss the API. :)

File: lib/iex/lib/iex/helpers.ex [hunk 1]
```diff
@@ -543,7 +555,10 @@ defmodule IEx.Helpers do
     case File.ls(path) do
       {:ok, items} ->
         sorted_items = Enum.sort(items)
-        ls_print(path, sorted_items)
+        print_function = fn(item, opts, width) -> 
+          format_item(Path.join(opts[:path], item), String.pad_trailing(item, width))
```
Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: We don't need to pass the opts argument. You can already access the `path` here as part of the closure.

File: lib/iex/lib/iex/helpers.ex [hunk 2]
```diff
@@ -560,19 +575,32 @@ defmodule IEx.Helpers do
...
+  First argument must be a list of string to print.
+
+  Second argument is optional and can be an anonymous function that format table cell.
+  Example: 
+      print_function = fn(item, opts, offset) -> String.pad_trailing(item, offset) end
+
+  Third argument is optional and can be a keyword list of values that can be used in format function
+  """
+  @spec tp([String.t], function, Keyword.t) :: any
+  def tp(list, print_function \\ fn(item, _opts, offset) -> String.pad_trailing(item, offset) end, opts \\ [])
```
Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: If we remove the opts argument from the anonymous function, which does not seem to be necessary, then the default print function can be `&String.pad_trailing/2`.

**Comment by Troush** [comment]
Okay :)

Made `tp` private again. And done all other fixes.

But I liked idea of public `tp` a lot. Better to create an Github Issue or a ML discussion about it?

File: lib/iex/lib/iex/helpers.ex [hunk 3]
```diff
@@ -534,6 +534,18 @@ defmodule IEx.Helpers do
   end
 
   @doc """
+  Produces a simple list of  all exports in a module
+  """
+  def e(module \\ Kernel) do
+    IEx.Autocomplete.exports(module) |> export_print()
+  end
+  
+  defp export_print(functions) do
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: It'd be better to name it `print_exports`, we usually tend to use such naming.

File: lib/iex/lib/iex/helpers.ex [hunk 4]
```diff
@@ -534,6 +534,18 @@ defmodule IEx.Helpers do
...
+    list = Enum.map(functions, fn({key, value}) -> Atom.to_string(key) <> "/" <> Integer.to_string(value) end)
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: I think `key` –> `name` and `value` –> `arity` will be more suitable.

File: lib/iex/lib/iex/helpers.ex [hunk 5]
```diff
@@ -560,19 +575,20 @@ defmodule IEx.Helpers do
 
   defp expand_home(other), do: other
 
-  defp ls_print(_, []) do
+  defp tp(list, print_function \\ &String.pad_trailing/2)
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Since we keep this function private is doesn't need to have such cryptic name, let's name it `print_table` then. :bowtie:

File: lib/iex/lib/iex/autocomplete.ex [hunk 6]
```diff
@@ -24,6 +24,14 @@ defmodule IEx.Autocomplete do
     end
   end
 
+  def exports(mod) do
```
Comment (lexmag) [inline lib/iex/lib/iex/autocomplete.ex]: I assume we should `@doc false` it and not test, right?

File: lib/iex/lib/iex/helpers.ex [hunk 7]
```diff
@@ -543,7 +555,10 @@ defmodule IEx.Helpers do
     case File.ls(path) do
       {:ok, items} ->
         sorted_items = Enum.sort(items)
-        ls_print(path, sorted_items)
+        print_function = fn(item, width) ->
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Inside`IEx.Introspection` module (and I believe other places as well) we call such function as `printer`, would be beneficial to be consistent I think.

**Comment by josevalim** [comment]
@Troush ML discussion would be best. Thank you.

File: lib/iex/lib/iex/autocomplete.ex (hunk 6 above)
Comment (josevalim) [inline lib/iex/lib/iex/autocomplete.ex]: Yes.

Comment (Troush) [inline lib/iex/lib/iex/autocomplete.ex]: If we are adding `@doc false` to `exports` do we maybe need add it to `expand` as well?

Comment (lexmag) [inline lib/iex/lib/iex/autocomplete.ex]: I think no, `expand/2` is the purpose and the main feature of the `IEx.Autocomplete` module, but `exports/1` is rather a shared helper function so we should emphasise it with `@doc false`.

**Comment by Troush** [comment]
All fixes done 👍

File: lib/iex/lib/iex/helpers.ex [hunk 8]
```diff
@@ -560,19 +575,20 @@ defmodule IEx.Helpers do
 
   defp expand_home(other), do: other
 
-  defp ls_print(_, []) do
+  defp table_print(list, printer \\ &String.pad_trailing/2)
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Let's name it `print_table` to follow convention. :bowtie:

**Comment by Troush** [comment]
💜

File: lib/iex/lib/iex/helpers.ex [hunk 9]
```diff
@@ -534,6 +534,18 @@ defmodule IEx.Helpers do
   end
 
   @doc """
+  Produces a simple list of  all exports in a module
```
Comment (whatyouhide) [inline lib/iex/lib/iex/helpers.ex]: Let's maybe rephrase as:

> Returns a list of all the functions and macros exported by the given module.

Wdyt? \cc @lexmag

**Comment by lexmag** [comment]
Thank you @Troush. 💛

File: lib/iex/lib/iex/helpers.ex (hunk 9 above)
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: @whatyouhide oops, merged already.

Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Seems we can fix similar wording for `ls` helper.

File: lib/iex/lib/iex/helpers.ex [hunk 10]
```diff
@@ -534,12 +534,12 @@ defmodule IEx.Helpers do
...
 
   @doc """
-  Produces a simple list of  all exports in a module
+  Produces a simple list of all exports in a module.
   """
   def e(module \\ Kernel) do
     IEx.Autocomplete.exports(module) |> print_exports()
   end
-  
+
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: @Troush seems your editor does not trim training whitespaces. :bowtie:
